performance_manager/lib/rt_vehicle_positions.py

```diff
@@ -0,0 +1,277 @@
+
+
+def start_time_to_seconds(
+    time: Optional[str],
+) -> Optional[float]:
+    """
+    transform time string in HH:MM:SS format to seconds integer
+    """
+    if time is None:
+        return time
+    (hour, min, sec) = time.split(":")
+    return int(hour) * 3600 + int(min) * 60 + int(sec)
+
+
+def transform_vp_dtyes(df: pandas.DataFrame) -> pandas.DataFrame:
+    """
+    ingest dataframe of vehicle position data from parquet file and transform
+    column datatypes
+    """
+    # current_staus: 1 = MOVING, 0 = STOPPED_AT
+    df["current_status"] = numpy.where(
+        df["current_status"] != "STOPPED_AT", 1, 0
+    ).astype(numpy.byte)
+
+    # store start_date as Int64 [nullable] instead of string
+    df["start_date"] = pandas.to_numeric(df["start_date"]).astype("Int64")
+
+    # store current_stop_sequence as Int64 [nullable]
+    df["current_stop_sequence"] = pandas.to_numeric(
+        df["current_stop_sequence"]
+    ).astype("Int64")
+
+    # store direction_id as Int64 [nullable]
+    df["direction_id"] = pandas.to_numeric(df["direction_id"]).astype("Int64")
+
+    # store start_time as seconds from start of day (Int64 [nullable])
+    df["start_time"] = (
+        df["start_time"].apply(start_time_to_seconds).astype("Int64")
+    )
+
+    # hash row data, exluding vehicle_timestamp column
+    # This hash is used to identify continuous series of records with the same
+    # categorical data
+    df["hash"] = df[list(set(df.columns) - {"vehicle_timestamp"})].apply(
+        lambda row: hash(tuple(row)), axis=1
+    )
+
+    # drop duplicates of hash and vehicle_timestamp columns (extraneous data)
+    df = df.drop_duplicates(subset=["vehicle_timestamp", "hash"])
+
+    return df
+
+
+def transform_vp_timestamps(df: pandas.DataFrame) -> pandas.DataFrame:
+    """
+    convert raw vehicle position data with one timestamp field into
+    vehicle position events with timestamp_start and timestamp_end
+    """
+    df = df.sort_values(by=["vehicle_id", "vehicle_timestamp"])
+
+    """
+    calculate mask to identify time series hash changes
+    this keeps first and last event for matching consecutive hash events
+    result is a 'start' event and 'end' event in consecutive rows for 
+    matching hash events
+    """
+    first_last_mask = (df["hash"] - df["hash"].shift(1) != 0) | (
+        df["hash"] - df["hash"].shift(-1) != 0
+    )
+    df = df.loc[first_last_mask, :]
+
+    """
+    transform vehicle_timestamp with matching consectuive hash events
+    into one record with timestamp_start and timestamp_end columns
+    if event is not part of consectuive matching events, then 
+    timestamp_start and timestamp_end will be same value
+    """
+    df.rename(columns={"vehicle_timestamp": "timestamp_start"}, inplace=True)
+    df["timestamp_end"] = numpy.where(
+        (df["hash"] - df["hash"].shift(-1) == 0),
+        df["timestamp_start"].shift(-1),
+        df["timestamp_start"],
+    ).astype("int64")
```

Comment:
if we've dropped all duplicate hashes above, won't this value always be non zero?

Reply:
The duplicates that are dropped in `transform_vp_dtyes` are duplicated where `hash` and `vehicle_timestamp` are both the same. After that drop, we could still have multiple records with the same `hash` and different `vehicle_timestamp` values.
 
Then, right before this operation, we do a drop that keeps only the first and last `vehicle_timestmap` for consecutive records with the same `hash`. This leaves is with, at most, two records for each consecutive run of similar `hash` values. 

This operation creates our `timestamp_end` value from the `timestamp_start` of the 2nd of any records having consecutive hashes.